src/misc.rs

```diff
@@ -47,6 +50,190 @@ where
         .unwrap_or_else(|| Vec::new())
 }
 
+/// Given as input `elements`, an array of field elements
+/// \\(\rho_0, \dots, \rho_{n-1}\\)
+/// compute the tensor product
+/// \\( \otimes_j (1, \rho_j )\\)
+pub fn tensor<F: Field>(elements: &[F]) -> Vec<F> {
+    assert!(!elements.is_empty());
+    let mut tensor = vec![F::one(); 1 << elements.len()];
+    let mut elements_iterator = elements.iter().enumerate();
+
+    tensor[1] = *elements_iterator
+        .next()
+        .expect("Expecting at least one element in the tensor product.")
+        .1;
+    // guaranteed to have at least one element.
+
+    for (i, element) in elements_iterator {
+        for j in 0..1 << i {
+            tensor[(1 << i) + j] = tensor[j] * element;
+        }
+    }
+    tensor
+}
+
+pub(crate) type PartialTensor<F> = Vec<Vec<F>>;
+
+/// Partially expand the tensor product
+/// \\(\otimes (1, \rho_j)\\)
+/// XXX TODO: This function is pub(crate) as in a previous version of this library,
+/// Iterable: Copy and hence couldn't store vectors itself.
+/// This is not anymore the case thus it can be moved inside init.
+pub fn expand_tensor<F: Field>(elements: &[F]) -> PartialTensor<F> {
```

Comment:
this is my code, why are you adding it here now? also what about XXX TODO? can you fix them?

Reply:
We talked about this on Wednesday - it appears as an addition due to moving around `multikzg` additions to `misc.rs` but should be unchanged from when you wrote it. It's used by `psnark` and `snark` and we are not sure if we should touch/edit this code to fix the TODO. What do you think?